add mobile part to project

codci opened the pull request:
* add mobile part to instruments part base on common webdriver solution
* add Mobile config to config_manager
* add project/test sample based on Yahoo weather app

todo list:
1) review/update mobile element class
2) change realization of BaseElements class
3) add pytest option to work with mobile devices / apps
4) update project documentation

disokol committed:
* add Appium local server class to remote_server.py
* add mobile part to instruments part base on common webdriver solution
* add Mobile config to config_manager
* add project/test sample based on Yahoo weather app

todo list:
1) review/update mobile element class
2) change realization of BaseElements class
3) add pytest option to work with mobile devices / apps
4) update project documentation

disokol committed:
* add Appium local server class to remote_server.py
* add mobile part to instruments part base on common webdriver solution
* add Mobile config to config_manager
* add project/test sample based on Yahoo weather app

todo list:
1) review/update mobile element class
2) change realization of BaseElements class
3) add pytest option to work with mobile devices / apps
4) update project documentation

disokol committed:
* add Appium local server class to remote_server.py
* add mobile part to instruments part base on common webdriver solution
* add Mobile config to config_manager
* add project/test sample based on Yahoo weather app

todo list:
1) review/update mobile element class
2) change realization of BaseElements class
3) add pytest option to work with mobile devices / apps
4) update project documentation

disokol committed:
* add Appium local server class to remote_server.py
* add mobile part to instruments part base on common webdriver solution
* add Mobile config to config_manager
* add project/test sample based on Yahoo weather app

todo list:
1) review/update mobile element class
2) change realization of BaseElements class
3) add pytest option to work with mobile devices / apps
4) update project documentation

disokol committed:
* add Appium local server class to remote_server.py
* add mobile part to instruments part base on common webdriver solution
* add Mobile config to config_manager
* add project/test sample based on Yahoo weather app

todo list:
1) review/update mobile element class
2) change realization of BaseElements class
3) add pytest option to work with mobile devices / apps
4) update project documentation

disokol committed:
* add Appium local server class to remote_server.py
* add mobile part to instruments part base on common webdriver solution
* add Mobile config to config_manager
* add project/test sample based on Yahoo weather app

todo list:
1) review/update mobile element class
2) change realization of BaseElements class
3) add pytest option to work with mobile devices / apps
4) update project documentation

maximrodin commented on common/_webdriver_qa_api/core/base_elements.py:
```diff
@@ -307,15 +314,16 @@ def assert_elements_number_greater_than(self, expected):
+    def get_elements(self):
+        """
+        find elements and get it's text
+        :return: list of the text of element
```
Incorrect docstring

codci replied:
fixed

maximrodin commented on common/_webdriver_qa_api/core/utils.py:
```diff
@@ -1,12 +1,10 @@
 import time
-import logging
 from datetime import timedelta
+from common.facade import logger
```
Please do not use a facade logger inside the core. Need to use built-in logger that will be replaced with framework logger during initialization

codci replied:
there was a problem in def fail_test(message): function. 

So we need to move it or use logger there. As fast fix I just import logger, but we can discuss another way to solve it

maximrodin replied:
What was the problem? It didn't work or PyCharm just highlighted that log_fail not found?
The current framework logging init process is:
1) Initialize all core modules together with the default python logger
2) Initialize facade that replaces all python loggers with framework logger dynamically in runtime

maximrodin commented on common/_webdriver_qa_api/mobile/mobile_page.py:
```diff
@@ -0,0 +1,245 @@
+        try:
+            self.driver.hide_keyboard()
+        except WebDriverException as e:
+            type(e)
```
Redudant?

codci replied:
removed

maximrodin commented on common/_webdriver_qa_api/web/remote_server.py:
```diff
@@ -23,7 +23,11 @@ def stop_server(self):
         """
-        logger.info(f"stop webdriver server on {self.address}:{self.port}")
+        logger.info(f"stop WebDriver server on {self.address}:{self.port}")
+        for _ in range(5):
```
What is the purpose of such additional 5 times check?

codci replied:
We check that there is no new processes that could be start from another test session

maximrodin commented on common/_webdriver_qa_api/web/remote_server.py:
```diff
@@ -39,6 +43,51 @@ def _get_current_sessions(self):
+            cmd = f"netstat -na | find \"{self.port}\""
+        out, err, rc = subprocess_send_command(cmd)
+        # todo: try on windows (and modify expected command)
+        return bool(out[0] and 'node' in out[0])
```
It Will not work on windows. How about to move is_local_server_running to BaseRemoteServer? Because the same logic is used for SeleniumServer class

codci replied:
Thats why I added todo section, let's resolve it on Windows firstly and next will think about moving to Base and override with specific server commands

maximrodin replied:
I've already replaced in with the function above in SeleniumServer)

    def is_local_server_running(self):
        """
        Try to get process info and return True if port now used.
        """
        if self.is_linux:
            cmd = f"lsof -i -n -P | grep {self.port}"
        else:
            cmd = f"netstat -na | find \"{self.port}\""
        out, err, rc = subprocess_send_command(cmd)
        return True if out[0] and ('LISTENING' or 'java') in out[0] else False

maximrodin commented on common/_webdriver_qa_api/mobile/mobile_driver.py:
```diff
@@ -0,0 +1,44 @@
+                capabilities.update(app=settings.ipa_path)
+                capabilities.update(udid=settings.ios_udid, deviceName=settings.ios_udid)
+            else:
+                raise ConfigError(f"Can't use iOS platform without 'ipa_path' and 'ios_udid' parameters.")
```
I suggest to move such validation from Driver directly to config to raise config error as soon as possible.
Place validation logic in _check_settings method of MobileSection class in config.
Like, define necessary parameters for android and ios platforms.

codci replied:
We can't be sure what platform will used and as result we can't be sure what parameters should be defined in config, second reason: we plan to add pytest option to override platform value and android / iOS parameters, that means that we will know about platforms after pytest options will be processed

maximrodin commented on common/_webdriver_qa_api/mobile/mobile_driver.py:
```diff
@@ -0,0 +1,44 @@
+                capabilities.update(deviceName=settings.android_udid, udid=settings.android_udid)
+                capabilities.update(appPackage=settings.android_package, appActivity=settings.android_activity)
+            else:
+                raise ConfigError(f"Can't use Android platform without "
```
I suggest to move such validation from Driver directly to config to raise config error as soon as possible.
Place validation logic in _check_settings method of MobileSection class in config.
Like, define necessary parameters for android and ios platforms.

codci replied:
same as comment before

maximrodin commented on config/pycats_config.ini:
```diff
@@ -10,12 +10,22 @@ validate_body = True
-selenium_server_executable = D:\Web Driver\selenium-server-standalone-3.141.59.jar
-chrome_driver_name = chromedriver.exe
+webdriver_folder = /Users/sdi/PycharmProjects/snoo-test-automation/framework/web_driver/common/
+selenium_server_executable = /Users/sdi/PycharmProjects/snoo-test-automation/framework/web_driver/common/selenium-server-standalone-3.14.0.jar
```
Please, do not update current config values that are already under git

codci replied:
fixed

codci replied:
@maximrodin 
probably we need to set some unused data instead of specified version of server. This should show that we need to fill in these fields before running tests?

maximrodin commented on project/tests/conftest.py:
```diff
@@ -51,3 +53,80 @@ def api_token():
+    request.addfinalizer(test_teardown)
+
+
+# def android_config(udid):
```
Can we remove unused code?

codci replied:
fixed

maximrodin commented on project/tests/test_mobile_example.py:
```diff
@@ -0,0 +1,33 @@
+
+
+@pytest.mark.usefixtures("start_mobile_session")
+@pytest.mark.mabile_sample
```
mobile_sample?

codci replied:
Yeap, it's one more way to run tests using custom mark

maximrodin replied:
No, I meant typo) mabile_sample != mobile_sample =)

disokol committed:
fix PR comments

disokol committed:
fix PR comments

disokol committed:
fix PR comments

maximrodin committed:
Fix Builder reinitialization to clean the structure after any update.
Replace string name of builder with __qualname__

maximrodin committed:
Fix Builder reinitialization to clean the structure after any update.
Replace string name of builder with __qualname__

maximrodin committed:
Fix Builder reinitialization to clean the structure after any update.
Replace string name of builder with __qualname__
Add unit test for builder init and fix unit tests for logging
Exclude Facade initialization for unit testing
Changed log level to INFO for API core

maximrodin committed:
Fix Builder reinitialization to clean the structure after any update.
Replace string name of builder with __qualname__
Add unit test for builder init and fix unit tests for logging
Exclude Facade initialization for unit testing
Changed log level to INFO for API core

maximrodin committed:
Merge pull request #18 from Softeq/bugfix/fix_api_model_init

Fix Builder reinitialization to clean the structure after any update.

disokol committed:
* add new web test sample (module test)
* update core / web webdriver part
* fix some codestyle problems

disokol committed:
* add new web test sample (module test)
* update core / web webdriver part
* fix some codestyle problems

disokol committed:
* add new web test sample (module test)
* update core / web webdriver part
* fix some codestyle problems

disokol committed:
* update webdriver metaclass to singleton
* fix some codestyle problems

disokol committed:
* update webdriver metaclass to singleton
* fix some codestyle problems

disokol committed:
* update webdriver metaclass to singleton
* fix some codestyle problems

disokol committed:
* move remote_server.py module to core part
* update mobile modules
* prepare structure to mobile module tests (sample)

disokol committed:
* move remote_server.py module to core part
* update mobile modules
* prepare structure to mobile module tests (sample)

disokol committed:
* move remote_server.py module to core part
* update mobile modules
* prepare structure to mobile module tests (sample)

disokol committed:
* move remote_server.py module to core part
* update mobile modules
* prepare structure to mobile module tests (sample)

disokol committed:
* update mobile_element.py / mobile_driver.py modules
* add switch.py module that expand mobile elements set
* add utils.py module with mobile platform actions
* update mobile project structure and test sample, add steps import flow that will used in tests

disokol committed:
* update mobile_element.py / mobile_driver.py modules
* add switch.py module that expand mobile elements set
* add utils.py module with mobile platform actions
* update mobile project structure and test sample, add steps import flow that will used in tests

disokol committed:
* update mobile_element.py / mobile_driver.py modules
* add switch.py module that expand mobile elements set
* add utils.py module with mobile platform actions
* update mobile project structure and test sample, add steps import flow that will used in tests

disokol committed:
* add AWSIotClient
* add PubnubClient
* add TestRailApiClient
* update logger (log_title method was added)
* update assertion functions
* migrated few web driver changes from iPic project

disokol committed:
* add AWSIotClient
* add PubnubClient
* add TestRailApiClient
* update logger (log_title method was added)
* update assertion functions
* migrated few web driver changes from iPic project

disokol committed:
* add AWSIotClient
* add PubnubClient
* add TestRailApiClient
* update logger (log_title method was added)
* update assertion functions
* migrated few web driver changes from iPic project

disokol committed:
* add AWSIotClient
* add PubnubClient
* add TestRailApiClient
* update logger (log_title method was added)
* update assertion functions
* migrated few web driver changes from iPic project

disokol committed:
* add AWSIotClient
* add PubnubClient
* add TestRailApiClient
* update logger (log_title method was added)
* update assertion functions
* migrated few web driver changes from iPic project

disokol committed:
* add AWSIotClient
* add PubnubClient
* add TestRailApiClient
* update logger (log_title method was added)
* update assertion functions
* migrated few web driver changes from iPic project

disokol committed:
* update logger (small changes in message format, new log_title method was added)
* WebDriver core part updates (few updates in WebElement class, assertions updated)
* support End-to-end test sample (fix test: add project classes with sample of custom element and page parts implementations)
* update module web tests

disokol committed:
* update logger (small changes in message format, new log_title method was added)
* WebDriver core part updates (few updates in WebElement class, assertions updated)
* support End-to-end test sample (fix test: add project classes with sample of custom element and page parts implementations)
* update module web tests

disokol committed:
* update logger (small changes in message format, new log_title method was added)
* WebDriver core part updates (few updates in WebElement class, assertions updated)
* support End-to-end test sample (fix test: add project classes with sample of custom element and page parts implementations)
* update module web tests

disokol committed:
* update logger (small changes in message format, new log_title method was added)
* WebDriver core part updates (few updates in WebElement class, assertions updated)
* support End-to-end test sample (fix test: add project classes with sample of custom element and page parts implementations)
* update module web tests

disokol committed:
* update logger (small changes in message format, new log_title method was added)
* WebDriver core part updates (few updates in WebElement class, assertions updated)
* support End-to-end test sample (fix test: add project classes with sample of custom element and page parts implementations)
* update module web tests

disokol committed:
fix PR comments

disokol committed:
fix PR comments

disokol committed:
fix PR comments

maximrodin committed:
Merge pull request #19 from Softeq/HB_migration

Hb migration - web part

disokol committed:
Merge branch 'develop' into hb_migration_utils

disokol committed:
* fix PR comments

disokol committed:
* fix PR comments

disokol committed:
fix PR changes

disokol committed:
fix PR changes

disokol committed:
update wait_element / try_wait_element

disokol committed:
update wait_element / try_wait_element

codci committed:
Merge pull request #21 from Softeq/hb_migration_utils

Hb migration: Web part updates + add new core modules

disokol committed:
* rebase to develop

disokol committed:
* rebase to develop

disokol committed:
Merge remote-tracking branch 'origin/mobile_module' into mobile_module

# Conflicts:
#	common/_libs/helpers/singleton.py
#	common/_webdriver_qa_api/core/base_elements.py
#	common/_webdriver_qa_api/core/text_box_mixin.py
#	common/_webdriver_qa_api/core/utils.py
#	common/_webdriver_qa_api/web/web_elements.py
#	project/tests/conftest.py
#	project/web/pages/main_page.py
#	requirements.txt

disokol committed:
update after merging from develop

disokol committed:
update after merging from develop

disokol committed:
Merge branch 'mobile_module' into mobile_module_migration

# Conflicts:
#	project/tests/conftest.py
#	project/tests/test_weather_example.py
#	project/tests/web_tests/test_sign_in_page.py

disokol committed:
PR fixes

disokol committed:
PR fixes

disokol committed:
PR fixes

disokol committed:
PR fixes

codci committed:
Merge pull request #20 from Softeq/mobile_module_migration

Mobile module updates

disokol committed:
* update WebDriver/MobileDriver classes (used DTO instead configParser)

codci committed:
Merge pull request #22 from Softeq/update_web_mobile_drivers

* update WebDriver/MobileDriver classes
